### ISSUE TITLE
[FEATURE REQ] Support TokenCredential in Azure Files SDK (ShareClient, etc.)

### ISSUE THREAD

**Issue by jstutson-rel** [Azure/azure-sdk-for-net#36310]
### Library name

Azure.Storage.Files.Share

### Please describe the feature.

Resubmitting closed feature request https://github.com/Azure/azure-sdk-for-net/issues/17000.

The ShareClient of Azure.Storage.Files.Share should support the DefaultAzureCredential.  Or another means of retrieving current size and quota for a file share should be provided that supports TokenCredential.

`    var client = new ShareClient(
        new Uri("https://myaccountname.file.core.windows.net/mysharename"),
        new DefaultAzureCredential());`

Look at the equivalent flow in powershell. No shared access key is needed.

```
#Connect to Azure and Import Az Module
Connect-AzAccount -Identity
Import-Module -Name Az.Accounts
Import-Module -Name Az.Storage
Set-AzContext -SubscriptionId $subscription_id

# Get file share
$StorageContext = New-AzStorageContext -StorageAccountName $storage_account_name -Anonymous
$PFS = Get-AzRmStorageShare -ResourceGroupName $resource_group -StorageAccountName $storage_account_name -Name $file_share_name -GetShareUsage

# Get provisioned capacity and used capacity
$ProvisionedCapacity = $PFS.QuotaGiB
$UsedCapacity = $PFS.ShareUsageBytes
```

The object returned by Get-AzRmStorageShare may be quite different from a ShareClient, but it does provide the size and quota information commonly needed for quota monitoring.  This is an extremely common need for Azure Files and it's already perplexing enough that this kind of auto-expansion logic isn't already built-in to the platform on top of needing to make two API calls to get the size and quota (GetStatisticsAsync and GetPropertiesAsync) vs the single line in Powershell.

The reason for not supporting TokenCredentials with ShareClient in the prior feature request was that SMB protocol doesn't support it.  Does getting size and quota actually require SMB protocol?  If not, it would be good to get a separate client that can be used to retrieve this information using the DefaultAzureCredential().  
If it does need SMB, then how is powershell pulling this off?  It would seem it's doing some work for us under the covers.

**Comment by PramodValavala-MSFT** [comment]
@jstutson-rel Thanks for the feedback! We have routed this to the owning team to review.

**Comment by jstutson-rel** [comment]
Hello, any word back on this?
It's actually looking potentially worse than I thought.  The share(s) Azure File Shares I am accessing all require virtual network peering to access.  This is really only meant to apply to actual file access, and again, I only need to get the current usage and size quota properties and update the quota if necessary.  The network peering really should not apply, and I don't think it does when accessing via AZ CLI. 

If I have to peer my function app with these shares just to get size and quota, I'm going to have to upgrade from the consumption plan to the premium plan.  This is not good.

**Comment by amnguye** [comment]
Hi,

Is the issue that this line of code isn't working for you? Because `ShareClient` and `ShareFileClient` do support `TokenCredential` currently (e.g. 12.14.0-beta.1 and above).
```
var client = new ShareClient( new Uri("https://myaccountname.file.core.windows.net/mysharename"), new DefaultAzureCredential());
```

I didn't see an attached line of code that says what API call you're making with the `client` that's returned. Could you return the stack trace that is occuring from the API, along with which API you're calling?

**Comment by jstutson-rel** [comment]
I am using Azure.Storage.Files.Shares v12.14.0

```
				Response<ShareStatistics> stats = await shareClient.GetStatisticsAsync().ConfigureAwait(false);
				shareUsageBytes = stats.Value.ShareUsageInBytes;
				int shareUsageGiB = (int)Math.Ceiling((double)shareUsageBytes / 1024 / 1024 / 1024);
				Response<ShareProperties> shareProperties = await shareClient.GetPropertiesAsync().ConfigureAwait(false);
				shareQuotaGiB = shareProperties.Value.QuotaInGB ?? 0;
```
As mentioned, I am retrieving share usage and share quota to ensure adequate headroom.  Are there other APIs that can retrieve this using ShareClient?

I will try using the pre-release version with a token credential and see if this avoids the need to be VNET peered and use the shared access key.  Would this pre-release version be safe to use across all regions of Azure?

**Comment by amnguye** [comment]
Yeah unfortunately looks like this is a limitation in the Files REST API (the SDK takes a dependency on the REST API, since it is used underneath).

But yes you would have to use the Azure Storage Resource Provider API instead. Which would be the [Microsoft.Azure.Management.Storage](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/storage/Microsoft.Azure.Management.Storage) instead.

@blueww Could you help out with the corresponding API they need to achieve setting the size and quota for a File Share (using Oauth).

**Comment by jstutson-rel** [comment]
I believe got it figured out with the move to the Azure Storage Resource Provider API.  Thanks for confirming.

I do have a couple concerns with this move however. 
1. I am curious if this API has the same or worse throttle limitations on calls per x seconds vs the APIs that ShareClient uses.  We might be pushing near this limit.
2. The API spec mentions that ShareUsageBytes property returned by it is only "approximate" and may not include most recent writes, but it gives no additional detail as to how up-to-date this information is.  Seconds?  Minutes?  Hours?  It's crucial that I know how frequently this is updated.

I'll be testing both out but very much appreciate any answers here.

**Comment by amnguye** [comment]
@blueww Could you help answer this question?

**Comment by blueww** [comment]
@jstutson-rel 
To get File share from storage management plan SDK, 
[Microsoft.Azure.Management.Storage](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/storage/Microsoft.Azure.Management.Storage)  is deprecated, please use [Azure.ResourceManager.Storage](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/storage/Azure.ResourceManager.Storage) instead.
You can refer [this sample](https://github.com/Azure/azure-sdk-for-net/blob/a805f13f7235bc0e36bbe9ff2797342d04170bc9/sdk/storage/Azure.ResourceManager.Storage/samples/Generated/Samples/Sample_FileShareCollection.cs#L278) for how to get File share usage.

For the 1st question of throttle limitations, you can refer this doc : https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/azure-subscription-service-limits#azure-storage-resource-provider-limits
This should be read operation, and the limit is "800 per 5 minutes" per region.

For the 2nd question, this is a question on server API behavior, as client side tools/SDK owner, we might not be the best person to answer it. I would suggest you to raise a help request in Azure portal like following picture, then the related team will be engaged to help you? (If you have concern for the throttle limitation, you can also raise it in the help request.)
![image](https://github.com/Azure/azure-sdk-for-net/assets/6926193/acd84d2d-6e9c-46d6-96dd-2c555cecf230)

**Comment by jstutson-rel** [comment]
Thanks for confirming the request quota on Resource Manager but I'm curious how that compares to the quota on the ShareClient.  Does anyone know?

In summary the optimal API for our file share monitoring has the following:
- A high request limit (800 per 5 minutes fits us for now but inhibits growth a bit)
- Up to the minute (or at absolute most 20 minute) accuracy of current usage bytes.
- OAuth/Token authentication with Azure AD (User Managed Identities)
- No VNet restriction

Looks like the resource manager version of these APIs will work well enough for us to monitor for adequate headroom and allocate more if needed from Azure Functions.  The request limit remains a concern and I'm awaiting response from support on the SLO for usage stats, but I think we can close this.

**Comment by ParmeetSingh** [comment]
I get this error actually while using above library.

Status: 400 (An HTTP header that's mandatory for this request is not specified.)
ErrorCode: MissingRequiredHeader

Additional Information:
HeaderName: x-ms-file-request-intent

Content:
﻿<?xml version="1.0" encoding="utf-8"?><Error><Code>MissingRequiredHeader</Code><Message>An HTTP header that's mandatory for this request is not specified.
